doc/CONFIGURATION.md
# Configuration

JLineup has many configuration options. A JLineup run with it's _before_ and _after_ steps is called _job_ throughout
this documentation. Job configuration settings can be separated between two scopes, _global_ and _site_. Some of the
options exist in global as in site scope. Site specific settings always have precedence before global settings. If you want to
have the same settings for all sites in your JLineup job, then you can simply specify those settings globally.

### Format

A JLineup config is a json document. The default name of the config file is ___lineup.json___.
The command line version looks for it in the current working directory. A different path can be specified with the
`--config` option when you call jlineup via command line.

### Basic config

A basic configuration with default settings can look like this:

```json
{
  "urls": {
    "https://www.otto.de": {
    }
  }
}
```
 
This config leads JLineup to https://www.otto.de. All other settings like window size, timeouts stay default.

What about those __global__ and __site__ settings?

```json
{
  "urls": {
    "https://www.otto.de": {
        "->": "SITE SETTINGS GO HERE"
    }
  },
  "->": "GLOBAL SETTINGS GO HERE"
}
```
    
### Example

This basic example shows the concept of global and url settings:

```json
{
  "urls": {
    "http://www.otto.de": {
      "paths": [
        "/"
      ],
      "window-widths": [
        1200
      ],
      "max-scroll-height":50000
    },
    "http://www.example.com": {
      "paths": [
        "/",
        "somepath"
      ],
      "window-widths": [
        800,
        1000
      ],
      "max-scroll-height":10000
    }
  },
  "browser": "chrome",
  "wait-after-page-load": 1
}
```
    
As you can see, _paths_, _window-widths_ and _max-scroll-height_ are configured on url level, the _browser_ and the _wait-after-page-load_
options are global for all sites in this job.

### Full configuration

This is a full configuration with example values:

```json
{
  "urls": {
    "http://www.example.com": {
      "paths": [
        "/",
        "someOtherPath"
      ],
      "max-diff": 0.0,
      "cookies": [
        {
          "name": "exampleCookieName",
          "value": "exampleValue",
          "domain": "http://www.example.com",
          "path": "/",
          "expiry": "1970-01-01T01:00:01+0100",
          "secure": true
        }
      ],
      "env-mapping": {
        "live": "www"
      },
      "local-storage": {
        "exampleLocalStorageKey": "value"
      },
      "session-storage": {
        "exampleSessionStorageKey": "value"
      },
      "window-widths": [
        600,
        800,
        1000
      ],
      "max-scroll-height": 100000,
      "wait-after-page-load": 0.0,
      "wait-after-scroll": 0.0,
      "wait-for-no-animation-after-scroll": 0.0,
      "warmup-browser-cache-time": 0.0,
      "wait-for-fonts-time": 0.0,
      "javascript": "console.log(\u0027This is JavaScript!\u0027)",
      "hide-images": false,
      "http-check": {
        "enabled": false
      },
      "max-color-diff-per-pixel": 1
    }
  },
  "browser": "PhantomJS",
  "wait-after-page-load": 0.0,
  "page-load-timeout": 120,
  "window-height": 800,
  "report-format": 2,
  "screenshot-retries": 0,
  "threads": 0,
  "timeout": 600,
  "debug": false,
  "log-to-file": false,
  "check-for-errors-in-log": true,
  "http-check": {
    "enabled": false
  }
}
```

---

## Reference

What are all those options about? Look here for details:

---

### `urls`

 JLineup job settings for one or multiple sites have to be configured as JSON subdocument here
 
 * Scope: Global
 * Type: JSON Document
 * Default: None - ***urls*** is a mandatory config option
 * Example: `"urls": {
                   "https://www.otto.de": {}
            }`
            
---
            
### `browser`

 Defines, which browser is used for the JLineup job. The chosen browser has to be installed on the used system.
 One exception is PhantomJS¹. If it's configured and not installed, JLineup downloads and uses it.
 
 *Advice*: PhantomJS shouldn't be used any more, because it lacks more and more features of modern web and it's not 
 maintained any more.¹
 
 JLineup downloads a webdriver, but it doesn't install a real browser during runtime! 
 
 * Scope: Global
 * Type: String
 * Possible Values: `PhantomJS`, `Chrome`, `Firefox`, `Chrome-Headless`, `Firefox-Headless`
 * Default: `"browser": "PhantomJS"`
 * Example: `"browser": "Chrome-Headless"`
 
 ¹) PhantomJS Development has been suspended. For more details go to https://github.com/ariya/phantomjs/issues/15344
 
---

### `paths`         
               
 These paths are appended to the current site url and screenshotted individually
                                 
 * Scope: Site
 * Type: List of Strings
 * Default: `[ "" ]`
 * Example: `"paths": [ "/", "someOtherPath" ]`

---
 
### `max-diff`

 This is the maximum accepted difference of a single screenshot of the current site run. If the difference is
 greater than the configured maximum, then the failing url + path + window-width combination AND the complete JLineup
 job are marked as erroneous.
 
 The difference of every single screenshot at every scrolling position is calculated as a percentage between 0% (0.0) 
 and 100% (1.0) which means that every pixel of the two compared images is different.
 
 *Advice:* This option should be handled with care. I strongly suggest a value of 0.0 (which is default) because every
 accepted pixel difference could be *any* unintended change - which should be looked at by a human being! A CI job can't
 decide what change is intended or not through simply accept a percentage of pixel differences. 
 
 * Scope: Site or Global
 * Type: Float
 * Default: `0.0`
 * Min: `0.0`
 * Max: `1.0`
 * Example: `"max-diff": 0.01`     
 
---
                 
### `cookies`

 A list of cookies that are set on the site

 * Scope: Site
 * Type: JSON Document
 * Default: `{}`
 * Example: `
            "cookies": [
               {
                 "name": "exampleCookieName",
                 "value": "exampleValue",
                 "domain": "http://www.example.com",
                 "path": "/",
                 "expiry": "1970-01-01T01:00:01+0100",
                 "secure": true
               }
            ]
            `

---
    
### `env-mapping`

 This is a convenience option that can replace parts of the domain that are specified in the config before making
 screenshots.
 
 This can be nice if you want to replace environments that are generated into your lineup.json that don't match the 
 real site url. One Example: At OTTO, the generated live step would result in a generated url like https://live.otto.de,
 which is in reality reachable under https://www.otto.de.   
                  
 * Scope: Site
 * Type: Map
 * Default: `{}`
 * Example: `
            "env-mapping": {
                "live": "www"
            }
            `
            
---

### `local-storage`

 Sets key value pairs to the local storage of the site
 * Scope: Site
 * Type: Map
 * Default: Empty
 * Example: `"local-storage": {
        "exampleLocalStorageKey": "exampleLocalStorageValue"
      }`
      
---
      
### `session-storage`

 Sets key value pairs to the session storage of the site
 * Scope: Site
 * Type: Map
 * Default: Empty
 * Example: "session-storage": {
        "exampleSessionStorageKey": "exampleSessionStorageValue"
      }`
      
---      
      
### `window-widths`
  
 Every path in the site config will be screenshotted in these given window-widths of the browser
* Scope: Site
* Type: List of integers
* Unit: Pixels
* Default: `[ 800 ]`
* Example: `"window-widths": [
        600,
        800,
        1000
      ]`
      
---      
      
### `max-scroll-height`

 When using Chrome or Firefox, JLineup scrolls the page like a normal browser user, because that's the way the web
 works today. Nearly no page fits the screen, and modern pages tend to grow to huge sizes. To not scroll endlessly on
 large pages, there is a maximum value.
 
 * Scope: Site
 * Type: Integer
 * Default: `50000`
 * Unit: Pixels
 * Example: `"max-scroll-height": 100000`
 
--- 
 
### `wait-after-page-load`

 On modern web pages with JavaScript and asynchronous loading mechanisms, it's hard to tell when a page is really 
 loaded completely. JLineup uses Selenium under the hood. Selenium's get() method blocks until the browser/page fires
 an onload event (files and images referenced in the html have been loaded, but there might be JS calls that load more
 stuff dynamically afterwards). This option waits for the specified amount of time after this onload event was fired.
 This value should be chosen carefully by observing the time your screenshotted page needs until it's complete.
 
 If your reports show half loaded pages, you should increase the value until this does not happen any more. If you set
 it too high, you waste time in your runs.
 
 * Scope: Site or Global
 * Type: Float
 * Default: 0
 * Unit: Seconds
 * Example: `"wait-after-page-load": 3.5`
 
 ---
 
### `wait-after-scroll`

 If your reports show half loaded artifacts in lower parts of the page, you should increase this value until this does
 not happen any more. If you set it too high, you increase the time for site scrolling drastically. As long as your page
 does not load content dynamically during scrolling, you can keep this very low or at default.
 
 * Scope: Site
 * Type: Float
 * Default: 0
 * Unit: Seconds
 * Example: `"wait-after-scroll": 1.1`
 
---
 
### `warmup-browser-cache-time`

 If this value is greater than 0, JLineup simply loads the page prior to the screenshot run and waits the specified
 amount of seconds to fill the browser cache with everything that can be cached during page load. It can be helpful
 to have a prewarmed cache while doing the first screenshots of a site in a run.

 * Scope: Site
 * Type: Float
 * Default: 0
 * Unit: Seconds
 * Example: `"warmup-browser-cache-time": 5`
 
--- 

### `wait-for-fonts-time`

 If your web fonts load too slow, there may be cases that your page is rendered without a default alternate font,
 which is replaced by the loaded font during a screenshot run. You can explicitly wait for font loading with this
 option.
 
 This should be rarely needed. 

 * Scope: Site
 * Type: Float
 * Default: 0
 * Unit: Seconds
 * Example: `"wait-for-fonts-time": 3`
 
--- 
 
### `javascript`

 This is a mighty option. You can specify and run any given JavaScript after loading the page and prior to making the
 screenshots. This can be used for various things, like throwing things out of the DOM by id or class, replacing images etc.
 Be creative but careful :D
 
 * Scope: Site
 * Type: String
 * Default: null
 * Example: `"javascript": "console.log('This is JavaScript!')"`
 
--- 

### `hide-images`

 This option can be used to hide all images on the page before making screenshots. This can be useful, if you only
 want to check the layout, but not the possibly changing content of pictures on a site.

 * Scope: Site
 * Type: Boolean
 * Default: false
 * Example: `"hide-images": true`
 
---

### `http-check`

 It may be required to check connectivity to a page by checking a http return code before making screenshots of it.
 Real browsers don't report a return code to Selenium, because Selenium looks at a page from a user perspective.
 JLineup can make this extra check for any desired http return codes.
 
 This can be helpful, if you want to fail a JLineup job if the page shows a 404. If the *before* and *after* run are
 done on a non-reachable page without a HTTP check, the job would return no error - both error pages are 
 visually identical.
 
 If `enabled` is set to true, all return codes that are not in the whitelist of allowed codes are regarded as failure.
 If you don't explicitly specify allowed codes in the http-check document, there is a default list of these accepted 
 HTTP return codes: `200,202,204,205,206,301,302,303,304,307,308`
 
 * Scope: Site or Global
 * Type: JSON Document
 * Default: `{ "enabled": false, allowed-codes: [ 200,202,204,205,206,301,302,303,304,307,308 ] }`
 * Example: `
              "http-check": {
                "enabled": true,
                "allowed-codes": [
                  200,
                  202,
                  204,
                  205,
                  206
                ]
              }
            `
            
---            
            
### `max-color-diff-per-pixel`

 For special cases (i.e. rare anti alias issues), you can allow an accepted difference on RGB pixel level. 
 To be precise, it should be called *max-color-diff-per-color-per-pixel*, because every RGB color is handled
 individually. A comparison of the pixel RGB(255,255,255) and RGB(253,254,255) would result in a difference
 of 3. If max color diff would be below 3 in this case, an error would be reported.
 
 It doesn't count how often in the picture a difference occurs. The check is reset after each
 pixel while going through the image.
 
 * Scope: Site
 * Type: Integer
 * Default: `0`
 * Example: `"max-color-diff-per-pixel": 1`
 
---

### `page-load-timeout`

 This value is passed through to Selenium for loading the pages. From the Selenium docs:
 
    Sets the amount of time to wait for a page load to complete before throwing an error.
    If the timeout is negative, page loads can be indefinite.
    
 This timeout can help to not wait indefinitely when there are connectivity or page loading problems.
    
 * Scope: Global
 * Type: Integer
 * Unit: Seconds
 * Default: `120`
 * Example: `"page-load-timeout": 180`
  
--- 

### `name`

 You can define a name for your specific JLineup job. The name is shown in the report and on the jobs overview
 page if you use the web version of JLineup.
 
 * Scope: Global
 * Type: String
 * Default: `null`
 * Example: `"name": "Cool name for my JLineup Job"`
 
--- 
 
### `window-height`

 This is the height of the browser window which JLineup uses.
 
 * Scope: Global
 * Type: Integer
 * Unit: Pixels
 * Default: `800`
 * Example: `"window-height": 1000`
 
---
 
### `report-format`

 This is a deprecated option to switch to an older JSON report format, which is not that useful. Shouldn't be used any
 more and will be removed in the future.
 
 * Scope: Global
 * Type: Integer
 * Range: `1` or `2`
 * Default: `2`
 * Example: `"report-format": 2`
 
---

### `screenshot-retries`

 __This option should be used with care. Better don't use it at all. It *may* help in a flaky environment, but it's better
 to fix the flakiness then to use this option.__
 
 JLineup makes screenshots for every permutation of `url` + `path` + `window-width`. Internally, this combination is 
 called a **screenshot context**. If you specify retries and there is any error during screenshotting one of those
 contexts, this context is retried until a maximum of specified retries is reached. 
   
 * Scope: Global
 * Type: Integer
 * Default: 0
 * Example: `"screenshot-retries": 2`
 
--- 

### `threads`

 If you specify multiple urls, paths or window widths in your config, JLineup can run multiple browser instances in 
 parallel to speed up the job. Keep in mind, that memory and CPU requirements increase when multiple browsers are opened
 at the same time.
 
 If you use the CLI version, this simply sets the number of threads. Although the default is `0`, JLineup uses at least 
 one thread, so `0` or `1` make no difference in that case.
 
 The `0` has another meaning for the web server version, because a maximum of allowed threads per job is configured in
 the web server properties. If the job config says `0` (which is default), the job uses the maximum of allowed
 threads. You can override this through setting an explicit value. If your job configuration value is
 greater than the maximum in the web server properties, the latter wins - so your job cannot use more threads than the
 server allows. Simply leave this configuration at default until you have special requirements for this particular job.
 
 * Scope: Global
 * Type: Integer
 * Default: `0`
 * Example: `"threads": 2`
 
--- 
 
### `timeout`

 This is a *global* and *hard* timeout for the whole JLineup job. If the job isn't finished after this time, it fails.
 This can help to end stalled jobs. Of course, a job shouldn't stall, but there may be rare cases
 (which should be fixed!) - this timeout can help to not have an indefinite wait for JLineup in your pipeline in those 
 cases.
 
 * Scope: Global
 * Type: Integer
 * Unit: Seconds
 * Default: `600`
 * Example: `"timeout": 300`
 
--- 
 
### debug

 This option is for CLI only. If you set this to true, JLineup logs on `DEBUG` level. Can be useful for the CLI version, if you track down strange
 behavior or bugs - or if you simply want to know which steps JLineup is doing in detail.
     
 * Scope: Global
 * Type: Boolean
 * Default: `false`
 * Example: `"debug": true`
 
--- 

### log-to-file

 This option is for CLI only. The job output is logged to the file `jlineup.log` in the current working directory if 
 specified.
 
 * Scope: Global
 * Type: Boolean
 * Default: `false`
 * Example: `"log-to-file": true`
 
--- 
 
### check-for-errors-in-log

 This option only works for Chrome/Chromium. If there is any error logged to the Chrome console and this setting is
 set to `true`, an error is raised for this screenshot and the run fails.
 
 * Scope: Global
 * Type: Boolean
 * Default: `false`
 * Example: `"check-for-errors-in-log": false`
 
 

HISTORY.md
# JLineup Release History

## Release 3.0.0
* Massively improved documentation
* chromedriver 2.45

## Release 3.0.0-rc7
* Revert because it didn't work: Ability to wait for images to be fully loaded before making screenshots
* hide all images using javascript once the url configuration `hideImages` is enabled

## Release 3.0.0-rc6
* Ability to wait for images to be fully loaded before making screenshots. 
Especially important when using progressive jpegs in page.
 
## Release 3.0.0-rc5
* chromedriver 2.42
* geckodriver 0.23.0
* Version info displayed in web frontend
* Fix cookie warning in httpcheck

## Release 3.0.0-rc2
* Fix pom artifactIds (this helps if you try to download JLineup via Maven/Gradle)

## Release 3.0.0-rc1
* Major Refactoring
* JLineup can run as web service now

## Release 2.4.1
* Enhancement: Add jobConfig option `check-for-errors-in-log` to toggle whether jlineup stops on errors in browser log. Default is `true`

## Release 2.4.0
* Improvement: Better handling of global timeout
* Dependency updates
* New project structure

## Release 2.3.5
* Bugfix: Don't crash when no webdriver was downloaded before

## Release 2.3.4
* New Selenium and Webdrivermanager versions

## Release 2.3.3
* Bugfix: Don't stop jlineup run when there are severe errors in chrome after initial loading of page

## Release 2.3.2
* Bugfix: Limit max file lenght for screenshot files

## Release 2.3.1
* Enhancement: New log-to-file option in jobConfig enables logging to a file in the working dir

## Release 2.3.0
* Enhancement: Animated gifs are not played in firefox.

## Release 2.2.1
* Bugfix: Exit correctly when url is not reachable
* Experimental support for headless firefox (needs firefox nightly >55)

## Release 2.2.0
* Add headless support without the need for Xvfb for Chrome >59 (use browser "chrome-headless" in jobConfig)

## Release 2.1.1
* Update to chromedriver 2.30 to support Chrome 59.x

## Release 2.1.0
* Feature: Introduce global page load timeout, defaults to 120 seconds
* Feature: Global retry option if some Exception occurs during takeScreenshots, defaults to 0

## Release 2.0.1
* Bugfix: Fix difference image link in html report
* Update Geckodriver to 0.16.1

## Release 2.0.0
* Breaking change: New json report format with summary. Use "report-format":1 to use legacy format for json report.
* Breaking change: Screenshots folder is now below report folder by default. HTML report is in report folder.
* Breaking change: JLineup returns error code 1 if report shows a difference
* Selenium upgrade to 3.4
* Support for Firefox 53.x, Chrome 58.x
* Geckodriver 0.16.0, Chromedriver 2.29

## Release 1.5.3
* Some exceptions lead to stalling browser sessions, now every exception is thrown

## Release 1.5.2
* Bugfix: Close open browsers and webdrivers when some Selenium exception is throwing

## Release 1.5.1
* Bugfix: Don't start multiple webdriver downloads when running in multiple threads

## Release 1.5.0
* Update Geckodriver for Firefox to 0.15.0
* Update Chromedriver for Chrome to 2.28

## Release 1.4.1
* Bugfix: Version pinning of webdriver didn't work because of a change in webdrivermanager
  (https://github.com/bonigarcia/webdrivermanager/commit/29c531266c78399f3999b246da479163d734bee8)

## Release 1.4.0
* Cache warmup is always done with the greatest horizontal resolution of an url jobConfig
* It's possible to specify session storage, similar to local storage
* Increasing waiting time bug fixed (long sleep phases when using multiple threads and many paths/urls)

## Release 1.3.3
* It's possible to specify debug mode in jobConfig now ("debug":true), additionally to command line param

## Release 1.3.2
* Bugfix: JLineup 1.3.1 was not making any screenshots when wait-for-fonts-time was not 0

## Release 1.3.1
* Bugfix: Use legacy report format as default (will change with 2.0.0)

## Release 1.3.0
* Ability to wait for webfonts to be loaded before making screenshots

## Release 1.2.5
* Log uncaught Exceptions in Threadpool

## Release 1.2.4
* New wait after scroll option

## Release 1.2.3
* Move mouse pointer to 0,0 before making screenshots -
 some mouseover effects could affect the screenshots otherwise

## Release 1.2.2
* Use close before quitting webdrivers

## Release 1.2.1
* Fix for browser startup shuffle time

## Release 1.2.0
* It's possible to use multiple threads and browser instances to speed up

## Release 1.1.2
* --print-jobConfig now shows full example configuration

## Release 1.1.1
* The report contains the JLineup version at the bottom

## Release 1.1.0
* You can specify custom JavaScript that is executed before page screenshots start
  (i.e. useful to remove parts that change dynamically through third party
  - like ads - from the page)

## Release 1.0.12
* Don't strip http or https from image filenames
* Sanitize filenames properly

## Release 1.0.11
* Report image links work if url contains ? and/or =

## Release 1.0.10
* Report image links work if url contains hash (#) characters

## Release 1.0.9
* Report is interactive now

## Release 1.0.8
* Switch to Selenium 3
* New --debug option to be more verbose

## Release 1.0.7
* Scroll to top via JavaScript after initial page load
 (new Chrome sometimes moves a re-opened page to the last known scroll position)


core/src/main/java/de/otto/jlineup/config/Config.java
package de.otto.jlineup.config;

import com.fasterxml.jackson.annotation.JsonAlias;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonParseException;
import com.google.gson.annotations.SerializedName;
import de.otto.jlineup.browser.Browser;

import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.*;

import static de.otto.jlineup.config.Cookie.COOKIE_TIME_FORMAT;

@JsonIgnoreProperties(ignoreUnknown = true)
public final class Config {

    public static final String LINEUP_CONFIG_DEFAULT_PATH = "./lineup.json";
    public static final String EXAMPLE_URL = "https://www.example.com";

    public static final int DEFAULT_WARMUP_BROWSER_CACHE_TIME = 0;
    public static final int DEFAULT_REPORT_FORMAT = 2;

    static final Browser.Type DEFAULT_BROWSER = Browser.Type.PHANTOMJS;
    static final float DEFAULT_MAX_DIFF = 0;
    public static final int DEFAULT_WINDOW_HEIGHT = 800;
    static final float DEFAULT_GLOBAL_WAIT_AFTER_PAGE_LOAD = 0f;
    public static final int DEFAULT_WINDOW_WIDTH = 800;
    static final String DEFAULT_PATH = "/";
    static final int DEFAULT_MAX_SCROLL_HEIGHT = 100000;
    static final int DEFAULT_WAIT_AFTER_PAGE_LOAD = 0;
    static final int DEFAULT_WAIT_AFTER_SCROLL = 0;
    static final int DEFAULT_WAIT_FOR_NO_ANIMATION_AFTER_SCROLL = 0;
    static final int DEFAULT_WAIT_FOR_FONTS_TIME = 0;
    static final int DEFAULT_THREADS = 1;
    static final int DEFAULT_PAGELOAD_TIMEOUT = 120;
    static final int DEFAULT_SCREENSHOT_RETRIES = 0;
    static final int DEFAULT_GLOBAL_TIMEOUT = 600;

    public final Map<String, UrlConfig> urls;
    public final Browser.Type browser;

    @SerializedName(value = "wait-after-page-load", alternate = "async-wait")
    @JsonProperty("wait-after-page-load")
    @JsonAlias({"async-wait"})
    public final Float globalWaitAfterPageLoad;
    @SerializedName(value = "page-load-timeout")
    @JsonProperty("page-load-timeout")
    public final int pageLoadTimeout;
    @SerializedName("window-height")
    @JsonProperty("window-height")
    public final Integer windowHeight;
    @SerializedName("report-format")
    @JsonProperty("report-format")
    public final Integer reportFormat;
    @SerializedName("screenshot-retries")
    @JsonProperty("screenshot-retries")
    public final int screenshotRetries;
    @SerializedName("threads")
    @JsonProperty("threads")
    public int threads;
    @SerializedName("timeout")
    @JsonProperty("timeout")
    public final int globalTimeout;
    @SerializedName("debug")
    @JsonProperty("debug")
    public final boolean debug;
    @SerializedName("log-to-file")
    @JsonProperty("log-to-file")
    public final boolean logToFile;

    private final static Gson gson = new GsonBuilder().setDateFormat(COOKIE_TIME_FORMAT).setPrettyPrinting().create();

    /* Used by GSON to set default values */
    public Config() {
        this(configBuilder());
    }

    private Config(Builder builder) {
        urls = builder.urls;
        browser = builder.browser;
        globalWaitAfterPageLoad = builder.globalWaitAfterPageLoad;
        pageLoadTimeout = builder.pageLoadTimeout;
        windowHeight = builder.windowHeight;
        threads = builder.threads;
        screenshotRetries = builder.screenshotRetries;
        reportFormat = builder.reportFormat;
        globalTimeout = builder.globalTimeout;
        debug = builder.debug;
        logToFile = builder.logToFile;
    }

    public static Config parse(String config) {
        Config parsedConfig = gson.fromJson(config, Config.class);
        if (parsedConfig == null) {
            throw new JsonParseException(String.format("Config is not valid: '%s'", config));
        } else if (parsedConfig.urls == null) {
            throw new JsonParseException("No urls in JLineup config.");
        }
        return parsedConfig;
    }

    public static String prettyPrint(Config config) {
        return gson.toJson(config);
    }

    @Override
    public String toString() {
        return "Config{" +
                "urls=" + urls +
                ", browser=" + browser +
                ", globalWaitAfterPageLoad=" + globalWaitAfterPageLoad +
                ", pageLoadTimeout=" + pageLoadTimeout +
                ", windowHeight=" + windowHeight +
                ", reportFormat=" + reportFormat +
                ", screenshotRetries=" + screenshotRetries +
                ", threads=" + threads +
                ", globalTimeout=" + globalTimeout +
                ", debug=" + debug +
                ", logToFile=" + logToFile +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Config config = (Config) o;

        if (pageLoadTimeout != config.pageLoadTimeout) return false;
        if (screenshotRetries != config.screenshotRetries) return false;
        if (threads != config.threads) return false;
        if (globalTimeout != config.globalTimeout) return false;
        if (debug != config.debug) return false;
        if (logToFile != config.logToFile) return false;
        if (urls != null ? !urls.equals(config.urls) : config.urls != null) return false;
        if (browser != config.browser) return false;
        if (globalWaitAfterPageLoad != null ? !globalWaitAfterPageLoad.equals(config.globalWaitAfterPageLoad) : config.globalWaitAfterPageLoad != null)
            return false;
        if (windowHeight != null ? !windowHeight.equals(config.windowHeight) : config.windowHeight != null)
            return false;
        return reportFormat != null ? reportFormat.equals(config.reportFormat) : config.reportFormat == null;
    }

    @Override
    public int hashCode() {
        int result = urls != null ? urls.hashCode() : 0;
        result = 31 * result + (browser != null ? browser.hashCode() : 0);
        result = 31 * result + (globalWaitAfterPageLoad != null ? globalWaitAfterPageLoad.hashCode() : 0);
        result = 31 * result + pageLoadTimeout;
        result = 31 * result + (windowHeight != null ? windowHeight.hashCode() : 0);
        result = 31 * result + (reportFormat != null ? reportFormat.hashCode() : 0);
        result = 31 * result + screenshotRetries;
        result = 31 * result + threads;
        result = 31 * result + globalTimeout;
        result = 31 * result + (debug ? 1 : 0);
        result = 31 * result + (logToFile ? 1 : 0);
        return result;
    }

    public static Config defaultConfig() {
        return defaultConfig(EXAMPLE_URL);
    }

    public static Config defaultConfig(String url) {
        return configBuilder().withUrls(ImmutableMap.of(url, new UrlConfig())).build();
    }

    public static Builder configBuilder() {
        return new Builder();
    }

    public static Config exampleConfig() {
        return configBuilder()
                .withUrls(ImmutableMap.of("http://www.example.com",
                        new UrlConfig(
                                ImmutableList.of("/","someOtherPath"),
                                DEFAULT_MAX_DIFF,
                                ImmutableList.of(
                                        new Cookie("exampleCookieName", "exampleValue", "http://www.example.com", "/", new Date(1000L), true)
                                ),
                                ImmutableMap.of("live", "www"),
                                ImmutableMap.of("exampleLocalStorageKey", "value"),
                                ImmutableMap.of("exampleSessionStorageKey", "value"),
                                ImmutableList.of(600,800,1000),
                                DEFAULT_MAX_SCROLL_HEIGHT,
                                DEFAULT_WAIT_AFTER_PAGE_LOAD,
                                DEFAULT_WAIT_AFTER_SCROLL,
                                DEFAULT_WAIT_FOR_NO_ANIMATION_AFTER_SCROLL,
                                DEFAULT_WARMUP_BROWSER_CACHE_TIME,
                                "console.log('This is JavaScript!')",
                                DEFAULT_WAIT_FOR_FONTS_TIME
                        )))
                .build();
    }

    public static Config readConfig(final String workingDir, final String configFileName) throws FileNotFoundException {
        List<String> searchPaths = new ArrayList<>();
        Path configFilePath = Paths.get(workingDir + "/" + configFileName);
        searchPaths.add(configFilePath.toString());
        if (!Files.exists(configFilePath)) {
            configFilePath = Paths.get(configFileName);
            searchPaths.add(configFilePath.toString());
            if (!Files.exists(configFilePath)) {
                configFilePath = Paths.get(LINEUP_CONFIG_DEFAULT_PATH);
                searchPaths.add(configFilePath.toString());
                if (!Files.exists(configFilePath)) {
                    String cwd = Paths.get(".").toAbsolutePath().normalize().toString();
                    throw new FileNotFoundException("Config file not found. Search locations were: " + Arrays.toString(searchPaths.toArray()) + " - working dir: " + cwd);
                }
            }
        }
        BufferedReader br = new BufferedReader(new FileReader(configFilePath.toString()));
        return gson.fromJson(br, Config.class);
    }

    public static final class Builder {
        private Map<String, UrlConfig> urls = null;
        private Browser.Type browser = DEFAULT_BROWSER;
        private float globalWaitAfterPageLoad = DEFAULT_GLOBAL_WAIT_AFTER_PAGE_LOAD;
        private int pageLoadTimeout = DEFAULT_PAGELOAD_TIMEOUT;
        private int windowHeight = DEFAULT_WINDOW_HEIGHT;
        private int reportFormat = DEFAULT_REPORT_FORMAT;
        private int screenshotRetries = DEFAULT_SCREENSHOT_RETRIES;
        private int threads = DEFAULT_THREADS;
        private int globalTimeout = DEFAULT_GLOBAL_TIMEOUT;
        private boolean debug = false;
        private boolean logToFile = false;

        private Builder() {
        }

        public Builder withUrls(Map<String, UrlConfig> val) {
            urls = val;
            return this;
        }

        public Builder withBrowser(Browser.Type val) {
            browser = val;
            return this;
        }

        public Builder withGlobalWaitAfterPageLoad(float val) {
            globalWaitAfterPageLoad = val;
            return this;
        }

        public Builder withPageLoadTimeout(int val) {
            pageLoadTimeout = val;
            return this;
        }

        public Builder withWindowHeight(int val) {
            windowHeight = val;
            return this;
        }

        public Builder withReportFormat(int val) {
            reportFormat = val;
            return this;
        }

        public Builder withScreenshotRetries(int val) {
            screenshotRetries = val;
            return this;
        }

        public Builder withThreads(int val) {
            threads = val;
            return this;
        }

        public Builder withDebug(boolean val) {
            debug = val;
            return this;
        }

        public Builder withLogToFile(boolean val) {
            logToFile = val;
            return this;
        }

        public Builder withGlobalTimeout(int val) {
            globalTimeout = val;
            return this;
        }

        public Config build() {
            return new Config(this);
        }
    }
}
